### PR TITLE
Classical Gram-Schmidt for block vectors

### PR DESCRIPTION
**Pull request by peterrum**
FYI @vovannikov

### CODE REVIEW

File: include/deal.II/lac/solver_gmres.h [hunk 1]
```diff
@@ -730,7 +731,108 @@ namespace internal
...
+
+
+    template <typename VectorType,
+              std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
+                nullptr>
+    const typename VectorType::BlockType &
+    block(const VectorType &vector, const unsigned int b)
+    {
+      return vector.block(b);
+    }
```
Comment (peterrum) [inline include/deal.II/lac/solver_gmres.h:829]: In a next step, I will unify these with those from https://github.com/dealii/dealii/issues/14233.

File: include/deal.II/lac/solver_gmres.h [hunk 2]
```diff
@@ -730,7 +731,109 @@ namespace internal
...
+
+
+
+    template <class VectorType,
+              std::enable_if_t<
+                !is_dealii_compatible_distributed_vector<VectorType>::value,
+                VectorType> * = nullptr>
     void
     Tvmult_add(const unsigned int dim,
```
Comment (kronbichler) [inline include/deal.II/lac/solver_gmres.h:838]: I realize we did not have those before, but would you mind adding a short comment about what this function does?

**Review by kronbichler — APPROVED** [review]
Nice, it was not a lot of additional work to add this.

**Comment by marcfehling** [comment]
Looks like you need to update the output file.
```
4246/4609 Test #4245: lac/solver_gmres_01.debug .............................................................***Failed   11.35 sec

***  End of file "/jenkins/workspace/dealii_PR-14605/tests/lac/solver_gmres_01.output" reached while trying to read line 90.
***  File "output" has more lines than file "/jenkins/workspace/dealii_PR-14605/tests/lac/solver_gmres_01.output",
***  line 90 is the last one read from file "output"

make[3]: *** [solver_gmres_01.debug/diff] Error 1
make[2]: *** [CMakeFiles/lac.solver_gmres_01.debug.test.dir/all] Error 2
make[1]: *** [CMakeFiles/lac.solver_gmres_01.debug.test.dir/rule] Error 2
make: *** [lac.solver_gmres_01.debug.test] Error 2
Test lac/solver_gmres_01.debug: DIFF
===============================   OUTPUT BEGIN  ===============================
Generating solver_gmres_01.debug/interrupt_guard.cc
Scanning dependencies of target lac.solver_gmres_01.debug
Building CXX object CMakeFiles/lac.solver_gmres_01.debug.dir/solver_gmres_01.cc.o
Building CXX object CMakeFiles/lac.solver_gmres_01.debug.dir/solver_gmres_01.debug/interrupt_guard.cc.o
[  0%] Linking CXX executable solver_gmres_01.debug/solver_gmres_01.debug
[100%] Built target lac.solver_gmres_01.debug
Scanning dependencies of target lac.solver_gmres_01.debug.test
[100%] Generating solver_gmres_01.debug/output
[100%] Generating solver_gmres_01.debug/diff
lac/solver_gmres_01.debug: BUILD successful.
lac/solver_gmres_01.debug: RUN successful.
lac/solver_gmres_01.debug: DIFF failed. ------ Source: /jenkins/workspace/dealii_PR-14605/tests/lac/solver_gmres_01.output
lac/solver_gmres_01.debug: DIFF failed. ------ Result: /home/dealii/build/tests/lac/solver_gmres_01.debug/output
Check /home/dealii/build/tests/lac/solver_gmres_01.debug/output /jenkins/workspace/dealii_PR-14605/tests/lac/solver_gmres_01.output
lac/solver_gmres_01.debug: DIFF failed. ------ Diff:   /home/dealii/build/tests/lac/solver_gmres_01.debug/failing_diff
lac/solver_gmres_01.debug: DIFF failed. ------ First 20 lines of numdiff output:
----------------
          <==
##90      ==> DEAL::Solve with PreconditionIdentity: 

+++  File "/jenkins/workspace/dealii_PR-14605/tests/lac/solver_gmres_01.output" differs from file "output"
CMakeFiles/lac.solver_gmres_01.debug.test.dir/build.make:64: recipe for target 'solver_gmres_01.debug/diff' failed
CMakeFiles/Makefile2:7233: recipe for target 'CMakeFiles/lac.solver_gmres_01.debug.test.dir/all' failed
CMakeFiles/Makefile2:7240: recipe for target 'CMakeFiles/lac.solver_gmres_01.debug.test.dir/rule' failed
Makefile:2950: recipe for target 'lac.solver_gmres_01.debug.test' failed

lac/solver_gmres_01.debug: ******    DIFF failed    *******

===============================    OUTPUT END   ===============================
Expected stage PASSED - aborting
CMake Error at /home/dealii/build/share/deal.II/scripts/run_test.cmake:144 (message):
  *** abort
```

**Comment by kronbichler** [comment]
It would be good to get this in. @peterrum can you take a look at the failing test?

**Comment by peterrum** [comment]
> It would be good to get this in. @peterrum can you take a look at the failing test?

I have updated the test!

**Pull request merged**